### PR TITLE
feat(common/models): Edit path derivation

### PR DESCRIPTION
**Pull request by jahorton**
One feature that the initial edit-distance implementation was lacking:  the ability to trace the path used to find the minimum edit distance.  While not strictly necessary for the core search algorithm, it's useful for diagnosing any issues that may arise when debugging, etc.

As it turns out, it looks like this algorithm will be quite useful for 'aligning' contexts as the user continues typing.  A specialized, hand-written version was looking to be tricky to hand-code properly due to the wide variance of possible control paths, so I bumped up the priority of this previously-missing feature.

Note that a pair of `'transpose-start'` and `'transpose-end'` constitute a _single unit_ of edit distance.  It took me a little bit to settle on this representation, as this is a bit odd... but it does allow for perfect replication of the edits without a need to specify indices within either of the input or match sequences.

### CODE REVIEW

File: common/predictive-text/worker/correction/classical-calculation.ts [hunk 1]
```diff
@@ -161,6 +163,108 @@ namespace correction {
...
+          expectedCost += count;
+        }
+        ops.push('transpose-end');
+
+        // Double-check our expectations.
+        if(this.getCostAt(transposeParent[0]-1, transposeParent[1]-1) != currentCost - expectedCost) {
+          ops = null;
+        }
+        parent = [transposeParent[0]-1, transposeParent[1]-1];
+      }
```
Comment (eddieantonio) [inline common/predictive-text/worker/correction/classical-calculation.ts:203]:
```suggestion
      let [lastInputIndex, lastMatchIndex] = = ClassicalDistanceCalculation.getTransposeParent(this, row, col);
      if(lastInputIndex >= 0 && lastMatchIndex >= 0) {
        // OK, a transposition source is quite possible.  Still need to do more vetting, to be sure.
        let expectedCost = 1;

        // This transposition includes either 'transpose-insert' or 'transpose-delete' operations.
        ops = ['transpose-start']; // always needs a 'start'.
        if(lastInputIndex != row-1) {
          let count = row - lastInputIndex - 1;
          ops = ops.concat( Array(count).fill('transpose-delete') );
          expectedCost += count;
        } else {
          let count = col - transposeParent[1] - 1;
          ops = ops.concat( Array(count).fill('transpose-insert') );
          expectedCost += count;
        }
        ops.push('transpose-end');

        // Double-check our expectations.
        if(this.getCostAt(lastInputIndex-1, lastMatchIndex-1) != currentCost - expectedCost) {
          ops = null;
        }
        parent = [lastInputIndex-1, lastMatchIndex-1];
      }
```

File: common/predictive-text/unit_tests/headless/edit-distance/classical-calculation.js [hunk 2]
```diff
@@ -456,4 +456,78 @@ describe('Classical Damerau-Levenshtein edit-distance calculation', function() {
...
+      td: 'transpose-delete'
+    }
+
+    it("'accomodate' -> 'accommodate'", function() {
+      let buffer = compute("accomodate", "accommodate", "InputThenMatch");
+  
+      let editSequence = [
+      //  a     c     c     o     m
+        op.m, op.m, op.m, op.m, op.m,
+      // *m    o      d     a     t    e
```
Comment (darcywong00) [inline common/predictive-text/unit_tests/headless/edit-distance/classical-calculation.js:478]: Do you mark +m for `op.i`?

Comment (jahorton) [reply common/predictive-text/unit_tests/headless/edit-distance/classical-calculation.js:478]: ... yeah, kinda realized it after the PR went up... and after I'd split branch into two versions (as the changes are _also_ in #3549).  I'm being a bit lazy, but I could adjust it to use the same `+m` syntax.

Comment (jahorton) [reply common/predictive-text/unit_tests/headless/edit-distance/classical-calculation.js:478]: So, as noted:

```suggestion
      // +m    o      d     a     t    e
```

**Comment by mcdurdin** [comment]
Build failed with: 
```
worker/correction/classical-calculation.ts(180,46): error TS1109: Expression expected.
```

**Comment by jahorton** [comment]
Looks like I should have more thoroughly proofread that big suggestion before committing it...

**Comment by eddieantonio** [comment]
SORRY :/

**Pull request merged**